pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java

```diff
@@ -423,31 +421,65 @@ public void brokerConnected(DirectProxyHandler directProxyHandler, CommandConnec
     // According to auth result, send newConnected or newAuthChallenge command.
     private void doAuthentication(AuthData clientData)
             throws Exception {
-        AuthData brokerData = authState.authenticate(clientData);
-        // authentication has completed, will send newConnected command.
-        if (authState.isComplete()) {
-            clientAuthRole = authState.getAuthRole();
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("[{}] Client successfully authenticated with {} role {}",
-                        remoteAddress, authMethod, clientAuthRole);
+        CompletableFuture<AuthData> authChallengeFuture = authState.authenticateAsync(clientData);
+        if (authChallengeFuture.isDone()) {
+            if (!authChallengeFuture.isCompletedExceptionally()) {
+                authChallengeSuccessCallback(authChallengeFuture.get());
+            } else {
+                try {
```

Comment:
what about

```
authChallengeFuture.exceptionally(e -> {
    authenticationFailedCallback(unwrapCompletionException(e);)
});
```

and probably you could consolidate the two branches in one single "whenComplete" block

```
authChallengeFuture.whenComplete( (result, error) -> {
      if (error != null) {
                   authenticationFailedCallback(unwrapCompletionException(error));
     } else {
          authChallengeSuccessCallback(result);
     }
}); 
```

as you did below

Reply:
I wrote it this way because most `AuthenticationProvider` implementations are synchronous, and I was optimizing for the case where the future will always be completed. The `whenComplete` method incurs additional object creation and adding to a queue before executing right away. If you think this is a premature optimization, I can collapse the branches of this `if`.

Reply:
I collapsed the logic.